Add new city Isidora

jmbenlloch opened the pull request:
City to compute corrected waveforms (CWF) and store them in an HDF5 file. There is one test included which check that Isidora runs and that the corrected waveforms is equal to the BLR one within 1%. One sample config file is also included.

jmbenlloch committed:
Add new city Isidora

jacg commented on invisible_cities/cities/isidora_test.py:
```diff
@@ -0,0 +1,124 @@
+            BLR = pmtblr[event]
+            CWF = pmtcwf[event]
+            for pmt in range(len(BLR)):
+                diff = abs(np.sum(BLR[pmt][4950:5150]) \
```
Perhaps it would be helpful to replace the 4950 and 5150 *magic numbers* with meaningfully named variables?

jacg commented on invisible_cities/cities/isidora.py:
```diff
@@ -0,0 +1,231 @@
+                        if n_events_tot >= nmax and nmax > -1:
+                            print('reached max nof of events (={})'
+                                  .format(nmax))
+                            break
```
What is the point of the `nmax > -1` check? Can `nmax` ever be less than 0? I imagine that `nmax == 0` (or even `nmax >= 0` could be taken to mean 'unlimited number of events' in which case this meaning  needs to be mentioned in the docstring ... and in which case the check should be

    nmax > 0 and n_events_tot >= nmax

which is expressed more pythonically (but perhaps less clearly from the motivation perspective) as 

    n_events_tot >= nmax > 0

jacg commented on invisible_cities/cities/isidora.py:
```diff
@@ -0,0 +1,231 @@
+
+            except Exception:
+                print('error')
+                raise
```
This pattern (printing 'error' whenever an exception comes out) is repeated throughout the cities. Is it really necessary? Does it really add anything? I have a suspicion that sometime, somewhere, someone who doesn't understand Python very well, put this in somewhere, and now it is being copied all over the place.

Can we identify the reason for this? If no good reason exists, could we please put a stop to this practice and excise all occurances from our code? (I'll do it in a separate PR

jacg replied:
Please see my [remove-error-print branch](https://github.com/jacg/IC/tree/remove-error-print). It seems that the only other place where this is done in the new IC is `Irene.run()`. @jjgomezcadenas Please review the changes in the aforementioned branch to see whether there is any point in keeping this noise (best discussed face-to-face) and whether the finally-based version is preferable.

jacg commented on invisible_cities/cities/isidora_test.py:
```diff
@@ -107,16 +103,19 @@ def test_isidora_run(config_tmpdir):
+                diff = abs(np.sum(BLR[pmt][peak_start:peak_end]) -
+                           np.sum(CWF[pmt][peak_start:peak_end]))
                 diff = 100 * diff / np.sum(pmtblr[event][pmt])
                 assert diff < eps
```
Do we really need a loop here? Can't this be expressed as an array operation and/or some numpy array assertion?

jacg commented on invisible_cities/cities/isidora.py:
```diff
@@ -0,0 +1,231 @@
+        n_events_tot = 0
+        # check the state of the machine
+        if not self.input_files:
+            raise IOError('input file list is empty')
```
I see that this habit of using `IOError` in these situations is spreading throughout IC. I don't think that that is appropriate in such caes. Consider that from Python 3.3 onwards, `IOError` is just a synonym for `OSError`, and I hope that it's pretty clear that this is not an OS error! I think it would be better if we had specific IC exceptions with the exact meanings that we want to convey, and that we use them in these situations.

Let's nip this in the bud before too many people copy this bad practice!

jmbenlloch committed:
Vectorize loops

jmbenlloch committed:
Start a collection of IC Exceptions in core

ICException is the base exception for IC. So far, it has one subclass:
NoInputFiles

jacg commented on invisible_cities/core/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+
+class NoInputFiles(Exception):
+    """ Input files list is not defined """
+    pass
```
Nice.

Could we please make this a subclass of `InvisibleCitiesException` or `ICException`.

Our exception hierarchy will require a but more thought, but this is fine for now.

jmbenlloch replied:
I have already updated it in the last commit.
